### PR TITLE
Add dev warning when type prop is missed for checkbox, radio or multi…

### PR DESCRIPTION
**Pull request by Andarist**
Simple PR improving DX by providing a helpful warning to the developer about his/her mistake.

### CODE REVIEW

**Comment by erikras** [comment]
The `warning` function already does the `env` check. You can remove that.

**Comment by erikras** [comment]
Can you add a test? Warnings are tested like [this](https://github.com/final-form/react-final-form/blob/master/src/Field.test.js#L79-L96).

**Comment by erikras** [comment]
This is a great PR. I wonder if this warning shouldn't be in `getValue.js` since that's where the issue it's addressing is?

**Comment by Andarist** [comment]
@erikras I was just commenting about env check, here it goes:

Each `warning` call should be wrapped with explicit `NODE_ENV` check for better minification, uglifyJS and other minifiers are not always able to dead code eliminate things, you can inspect your minified umd bundle and you will see that log messages are still there

let's consider even really simple example of
```js
// input.js
const warning =
  'production' === 'production'
    ? (condition, message) => {}
    : (condition, message) => {
        if (!condition) {
          console.error(`Warning: ${message}`)
        }
      }

warning(false, 'Log')
```

```js
// uglifyjs input.js -c --toplevel
((condition,message)=>{})(!1,"Log");
```

I've also explained this lately in those 2 comments ([here](https://github.com/dollarshaveclub/postmate/pull/55#discussion_r171670788) && [here](https://github.com/dollarshaveclub/postmate/pull/55#issuecomment-370606874)) if you want to read slightly more.

My recommendation is to wrap each call site to remove all this dead code. I can do it as part of this PR or I can remove this "explicit" env check from this PR and prepare a follow up PR wrapping each call site.

> Can you add a test? Warnings are tested like this.

Gonna look into it in a moment and ping u when I'm done with it.

> I wonder if this warning shouldn't be in getValue.js since that's where the issue it's addressing is?

I've briefly thought about it, but `getValue` doesn't have access to `props` and it was easier to check this and construct more helpful message here. Also this problem affects `render` first but we cannot warn there because we don't know what user will render and to what kind of DOM element it will hook the handlers.

> This is a great PR.

Thanks! I thought you are gonna like it :wink:

**Comment by erikras** [comment]
> Each `warning` call should be wrapped with explicit `NODE_ENV` check for better minification

Awesome, I love being proved wrong! Makes total sense. I definitely wouldn't mind a PR converting all the other usages of `warning`...hint...hint... 🤔 😃

**Comment by erikras** [comment]
...actually, now that I think about it, this revelation kind of renders the `warning` function totally useless. 😆

**Comment by SimenB** [comment]
This is great! We were unable to get radios working properly with a custom component before sending in `type` to `Field`. Awesome with a great warning 🙂

File: src/Field.js [hunk 1]
```diff
@@ -100,6 +100,29 @@ export default class Field extends React.Component<Props, State> {
...
+
+      if (process.env.NODE_ENV !== 'production' && event && event.target) {
+        const targetType = event.target.type
+        const props: Object = this.props
+        const unknown =
+          ~['checkbox', 'radio', 'select-multiple'].indexOf(targetType) &&
+          !props.type
+
+        const type = targetType === 'select-multiple' ? 'select' : targetType
+        const { value }: any =
```
Comment (Andarist) [inline src/Field.js]: aint sure if this is entirely correct, if not - let me know what should be changed about it

File: src/Field.js [hunk 2]
```diff
@@ -151,8 +174,7 @@ export default class Field extends React.Component<Props, State> {
     } else if ((rest: Object).type === 'radio') {
       input.checked = value === _value
       input.value = _value
-    }
-    if (component === 'select' && (rest: Object).multiple) {
+    } else if (component === 'select' && (rest: Object).multiple) {
```
Comment (Andarist) [inline src/Field.js]: I think those are exclusive things, right?

File: src/Field.test.js [hunk 3]
```diff
@@ -706,4 +706,68 @@ describe('Field', () => {
...
+    TestUtils.Simulate.change(radio, {
+      target: { type: 'radio', value: 'radio_value' }
+    })
+    TestUtils.Simulate.change(select, {
+      target: { type: 'select-multiple', value: ['select_value'] }
+    })
+
+    expect(spy).toHaveBeenCalled()
+    expect(spy).toHaveBeenCalledTimes(3)
+    expect(spy.mock.calls[0][0]).toBe(
```
Comment (Andarist) [inline src/Field.test.js]: I find snapshots really nice for those kind of tests ("warning" text/errors thrown snapshots), but you do not use any snapshots so I wasn't sure if you'd like this approach.

**Comment by Andarist** [comment]
@erikras provided a test, please take a look :)

> Awesome, I love being proved wrong! Makes total sense. I definitely wouldn't mind a PR converting all the other usages of warning...hint...hint... 🤔 😃

Sure thing, gonna send additional PR in a minute.

> ...actually, now that I think about it, this revelation kind of renders the warning function totally useless. 

It still encapsulates branching logic, not much - but might be nice to keep it. We'll discuss this under upcoming PR.

File: src/Field.js (hunk 2 above)
Comment (erikras) [reply src/Field.js]: Hopefully no one wants a `<select type="radio">`. 🤣

File: src/Field.test.js (hunk 3 above)
Comment (erikras) [reply src/Field.test.js]: I only just recently understood snapshot testing. Not actively against it, but not excited to migrate any tests that could be more concise with snapshots... 🤷‍♂️

**Comment by erikras** [comment]
Published in [`v3.2.1`](https://github.com/final-form/react-final-form/releases/tag/v3.2.1).

**Pull request merged**